Add retries to some netlink calls

adrianmoisey opened the pull request:
#### What type of PR is this?

/kind bug
/kind flake

#### What this PR does / why we need it:

Some netlink calls fail with partial results. These can be retried.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/129562

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

aojea commented on pkg/proxy/util/utils.go:
```diff
@@ -223,3 +229,13 @@ func IsVIPMode(ing v1.LoadBalancerIngress) bool {
+			return
+		}
+	}
+	klog.V(3).InfoS("netlink call interrupted", "attempts", MaxAttemptsEINTR)
```
```suggestion
	klog.V(2).InfoS("netlink call interrupted", "attempts", MaxAttemptsEINTR)
```

adrianmoisey replied:
Fixed in f9fcbaecbe02b08961c4caa496768eb0c6646346

aojea commented:
/assign @aroradaman

aroradaman commented on pkg/proxy/util/nfacct/nfacct_linux_test.go:
```diff
@@ -478,7 +479,12 @@ func TestRunner_List(t *testing.T) {
-	counters, err := rnr.List()
+	var counters []*Counter
+
+	util.RetryOnIntr(func() error {
```
We don't need to handle this here you have already handled this in `nfacct_linux.go`. 
We can actually mock `unix.EINTR` in the `TestRunner_List` and `TestRunner_Get`.

For this test we can populate fake handler with mock data for two requests, first one should error out with `unix.EINTR` and second one should succeed and return the mock response. 

Since we are retrying on `unix.EINTR` we should not expect any error from `rnr.List()` here, we can also assert on the number of calls made.

adrianmoisey replied:
> We don't need to handle this here you have already handled this in `nfacct_linux.go`.

`nfacct_linux.go` is just bubbling the error up, hence the retry being handled here. Should I move the retry over to `nfacct_linux.go` ?

> For this test we can populate fake handler with mock data for two requests, first one should error out with unix.EINTR and second one should succeed and return the mock response.

I'll give that a go, thanks

aroradaman replied:
Yes, the retry logic needs to go to `nfacct_linux.go`. I assumed it is already there, sorry for the confusion.

adrianmoisey replied:
> We can actually mock `unix.EINTR` in the `TestRunner_List` and `TestRunner_Get`.

Just want to double check this.
From my understanding, List() makes sense to mock, see: https://github.com/kubernetes/kubernetes/blob/1ee2d3ca71f2f79c325ee92902fb92ea607b014c/pkg/proxy/util/nfacct/nfacct_linux.go#L149-L150
It is using `unix.NLM_F_DUMP`, so a partial response could be returned.

As for Get(): 
https://github.com/kubernetes/kubernetes/blob/1ee2d3ca71f2f79c325ee92902fb92ea607b014c/pkg/proxy/util/nfacct/nfacct_linux.go#L130-L131

It's not using a flag that may result in a partial result, so I assume there's no need to do anything there.

Does my logic make sense here?

aroradaman commented on pkg/proxy/util/utils.go:
```diff
@@ -223,3 +229,13 @@ func IsVIPMode(ing v1.LoadBalancerIngress) bool {
 	return *ing.IPMode == v1.LoadBalancerIPModeVIP
 }
+
+func RetryOnIntr(f func() error) {
```
I think we start a linux only file.

danwinship commented on pkg/proxy/util/utils.go:
```diff
@@ -223,3 +229,13 @@ func IsVIPMode(ing v1.LoadBalancerIngress) bool {
 	return *ing.IPMode == v1.LoadBalancerIPModeVIP
 }
+
+func RetryOnIntr(f func() error) {
```
There's already `retry.OnError` from `"k8s.io/client-go/util/retry"`. You should use that instead. (And then you can put the `retriable` helper func for it that compares against `EINTR` in the conntrack code so you don't have to worry about windows.)

adrianmoisey replied:
Thanks for that, I switched to it

danwinship commented on pkg/proxy/conntrack/conntrack.go:
```diff
@@ -57,8 +58,12 @@ func newConntracker(handler netlinkHandler) Interface {
+		entries, err = ct.handler.ConntrackTableList(netlink.ConntrackTable, netlink.InetFamily(ipFamily))
+		return err
+	})
+	return entries, err
```
So in the `EINTR` case, the list function still potentially returns a _partial_ list.

For `CleanStaleEntries`, if it gets back `EINTR` and a partial list from `ct.ListEntries`, it can still clean up any stale entries that appeared in that partial list, which is a better result than if it just throws the partial list away. So you should update `CleanStaleEntries` so if it gets back `EINTR` here, it still logs an error, but then continues trying to do cleanup.

(You should also update the `ListEntries` doc comment to mention that it can return a partial list and an error.)

adrianmoisey replied:
Gotcha, I've handled that. I'll double check if this applies to other places too

danwinship commented on pkg/proxy/util/nfacct/nfacct_linux.go:
```diff
@@ -160,7 +160,7 @@ func (r *runner) List() ([]*Counter, error) {
 		counters = append(counters, counter)
 	}
-	return counters, nil
+	return counters, err
```
You should update the `nfacct.Interface` docs to mention that it may return a partial list with an error.

danwinship reviewed (commented):
Please squash your fixup commits.

adrianmoisey commented:
> Please squash your fixup commits.

I plan to do this at the end, I'm finding the history quite useful at the moment as I figure this one out

adrianmoisey commented on pkg/proxy/util/nfacct/nfacct_linux.go:
```diff
@@ -146,9 +148,17 @@ func (r *runner) Get(name string) (*Counter, error) {
+
 	req := r.handler.newRequest(cmdGet, unix.NLM_F_REQUEST|unix.NLM_F_DUMP)
-	msgs, err := req.Execute(unix.NETLINK_NETFILTER, 0)
-	if err != nil {
+	err = retry.OnError(util.MaxAttemptsEINTR, util.ShouldRetryOnEINTR, func() error {
+		msgs, err = req.Execute(unix.NETLINK_NETFILTER, 0)
```
Should the `r.handler.newRequest(cmdGet, unix.NLM_F_REQUEST|unix.NLM_F_DUMP)` be included in the retry?
I only moved it out as the current mocking of the tests expects that.

adrianmoisey commented on pkg/proxy/conntrack/conntrack.go:
```diff
@@ -69,7 +75,13 @@ func (ct *conntracker) ClearEntries(ipFamily uint8, filters ...netlink.CustomCon
 	}
 
-	n, err := ct.handler.ConntrackDeleteFilters(netlink.ConntrackTable, netlink.InetFamily(ipFamily), filters...)
+	var n uint
+	var err error
+	err = retry.OnError(util.MaxAttemptsEINTR, util.ShouldRetryOnEINTR, func() error {
+		n, err = ct.handler.ConntrackDeleteFilters(netlink.ConntrackTable, netlink.InetFamily(ipFamily), filters...)
+		return err
+	})
```
🤔 
I'm unsure if this is correct. Since this is deleting entries, I assume the partial result returned is a list of deleted entries.
Retrying here until an error isn't returned means that the logs that are created below this retry, and over at https://github.com/kubernetes/kubernetes/blob/b828fc7701b5b25974fcb230f5bdd59318be58bf/pkg/proxy/conntrack/cleanup.go#L123-L127 could be incorrect

Should the retry logic be moved elsewhere? Or something else be changed here?

adrianmoisey replied:
My worry is a scenario like this...
Assume we have 100 entries that need deleting.
1. On first try, 20 are deleted, and the partial result error is returned
2. We retry
3. On second try, the remaining 80 are deletes, and no error is returned
4. The number of entries (80) is passed back to the caller, and the log says `"entriesDeleted", 80`, which is incorrect, since a total of 100 were deleted

I guess either the tally of records can be summed during the retry, or the retry logic can move.

adrianmoisey commented:
Hey @aojea @danwinship @aroradaman 

I've made a bunch of changes, I think it's getting close to be actually ready. I have a few open questions that I need help with

adrianmoisey commented on pkg/proxy/conntrack/cleanup.go:
```diff
@@ -42,8 +43,12 @@ func CleanStaleEntries(ct Interface, ipFamily v1.IPFamily,
 	if err != nil {
-		klog.ErrorS(err, "Failed to list conntrack entries")
-		return
+		if errors.Is(err, unix.EINTR) {
```
ListEntries itself has a retry, over here: 
https://github.com/kubernetes/kubernetes/blob/b828fc7701b5b25974fcb230f5bdd59318be58bf/pkg/proxy/conntrack/conntrack.go#L61-L67

I assume we don't want to retry in multiple places.

adrianmoisey commented on pkg/proxy/conntrack/fake.go:
```diff
@@ -26,6 +26,7 @@ import (
 // FakeInterface implements Interface by just recording entries that have been cleared.
 type FakeInterface struct {
 	entries []*netlink.ConntrackFlow
+	error   error
```
I took a look at doing this, and I'm not too sure it makes sense to do.

`CleanStaleEntries()` calls `ct.ListEntries` (the mocked function).

https://github.com/kubernetes/kubernetes/blob/b828fc7701b5b25974fcb230f5bdd59318be58bf/pkg/proxy/conntrack/cleanup.go#L38-L52

When CleanStaleEntries receives the interrupt error from `ct.ListEntries`, it just logs that and carries on.
From what I understand, the following test cases should be handled:
1. no error received - happy path
2. interrupt error received - Happy path but log that the interrupt was received
3. any other error received - return nothing

Since the retry logic is in the mocked function `ct.ListEntries`, no retrying will happen here.

Sorry if I'm misunderstanding you.

aroradaman replied:
> CleanStaleEntries() calls ct.ListEntries (the mocked function).

`ct.ListEntries` is not the mocked function here, we only mock `ct.handler.`
Assuming the case where the actual netlink handler returned interrupt, retry logic will be called, and if the second attempt passes, it will return the complete list with no error.  This is the goal of this task.

To mock this behavior we can use the fake handler which returns a partial list and throws an interrupt error on the first call, and a complete list with no error on the second call.

https://github.com/kubernetes/kubernetes/blob/b828fc7701b5b25974fcb230f5bdd59318be58bf/pkg/proxy/conntrack/conntrack.go#L61-L68


Note: We leverage the `netlinkHandler` interface for mocking, not the `Interface` interface.
https://github.com/kubernetes/kubernetes/blob/ed9572d9c7733602de43979caf886fd4092a7b0f/pkg/proxy/conntrack/conntrack.go#L30-L42

aojea replied:
I would not overcomplicate the testing on this, the retry function is a common pattern implemented across the repo, if we can make it simply great, if not 🤷

aroradaman replied:
There are two fake implementations currently, one for the **Interface** interface and other for **netlinkHandler** interface, I think we should consolidate that.
That will make this more easier.

aroradaman replied:
@adrianmoisey you can consider using this to consolidate the fake implementations here, 
https://github.com/kubernetes/kubernetes/commit/c86c9df71c9f052969f5e5c1f70928c98b51eea3

adrianmoisey replied:
I'm keen to try @aroradaman's suggestion to see if the tests can be nicer, if you're fine with that @aojea

aojea replied:
you are both driving this so I let you judge, I'm just skimming through the comments and trying to understand if we are reaching a point where "Perfect is the enemy of done"  :)

adrianmoisey replied:
Great. It's a good call out, but I feel like we're pretty close to the end. If this plan doesn't work out, I'll tidy up what I have and submit it for a review

aojea replied:
as I said, you both are driving the PR, we trust you

aojea commented:
squash and final review? it seems there is no much else to discuss

aroradaman committed:
pkg/proxy/conntrack: consolidate fake implementations

This commit consolidates the two fake implementations
*conntrack.Interface* and *conntrack.netlinkHandler*.
Since the handler of the *conntracker* can be mocked
using *conntrack.netlinkHandler* we don't need the
*FakeInterface* anymore.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

adrianmoisey committed:
Add retries to conntracker.ListEntries()

Signed-off-by: Daman Arora <aroradaman@gmail.com>

adrianmoisey committed:
Add retries to conntracker.ClearEntries()

Signed-off-by: Daman Arora <aroradaman@gmail.com>

adrianmoisey committed:
Add retries to runner.List()

adrianmoisey commented:
I've pushed my latest work. It's been rebased and split up into a commit per addition of retry, along with its test.
I've also refactored the tests according to @aroradaman's suggestions. I still need to add a test for runner.List() though

adrianmoisey commented:
/retest

adrianmoisey commented:
I think this may be ready now...

aojea commented:
I let @aroradaman do the review for LGTM

aroradaman commented on pkg/proxy/conntrack/fake.go:
```diff
@@ -28,12 +28,17 @@ type fakeHandler struct {
-	return fake.entries, nil
+	calls := fake.calls
+	fake.calls++
+	return fake.entries[calls], fake.errors[calls]
```
Can we handle the index out of range here in this commit instead of introducing that in https://github.com/kubernetes/kubernetes/pull/129704/commits/8b7249117956e9ef4be99a0c4cc30feb5c926698. Breaking the `calls` to `callsConntrackTableList` and `callsConntrackDeleteFilters` is fine for that commit but this commit should also handle the index error.

reason: this commit will fail the unit tests when fake proxiers calls for conntrack cleanup. IMHO, all individual commits of a PR should pass the CI.


And I'm not fully convinced with the title for this commit, as the retry logic is added in subsequent commits, not this commit.

aroradaman reviewed (commented):
The flow looks good, just some cosmetic changes and code movement around different commits is required.

Also bumping netlink (https://github.com/kubernetes/kubernetes/pull/129704/commits/f8c7f3a6a67f408b49a79db9b7784aac6ddfce4b)  should be the last commit of the PR instead of first.

adrianmoisey commented:
> Also bumping netlink ([f8c7f3a](https://github.com/kubernetes/kubernetes/commit/f8c7f3a6a67f408b49a79db9b7784aac6ddfce4b)) should be the last commit of the PR instead of first.

Done!

adrianmoisey commented:
/retest

aroradaman commented:
/lgtm

adrianmoisey commented:
/assign danwinship
/assign aojea

aojea commented on pkg/proxy/conntrack/cleanup_test.go:
```diff
@@ -276,6 +276,11 @@ func TestCleanStaleEntries(t *testing.T) {
 	fake := &fakeHandler{}
 	fake.entries = mockEntries
+	// we will simulate 3 calls to ConntrackTableList:
+	// 1. Partial result + EINTR - first list call - retried
+	// 2. Full result - second list call - not retried
+	// 3. Full result - for the subsequent list call
```
I do not understand why is there a third call if the 2nd succeeds?

aojea commented on pkg/proxy/conntrack/fake.go:
```diff
@@ -20,20 +20,53 @@ limitations under the License.
+	tableListErrors []error
+
+	// callsConntrackTableList is a counter that gets incremented each time ConntrackTableList is called
+	callsConntrackTableList int
```
you can reduce the `tableListErrors`  as it is a queue, so you don't need the counter

aojea commented on pkg/proxy/conntrack/fake.go:
```diff
@@ -20,20 +20,53 @@ limitations under the License.
+	}
+
+	calls := fake.callsConntrackTableList
+	fake.callsConntrackTableList++
+
+	// we panic when ConntrackTableList() is called more times than the length of configured
+	// tableListErrors. This indicates improper configuration of fakeHandler for the test.
+	if calls > len(fake.tableListErrors)-1 {
+		panic("exceeded the allowed number of ConntrackTableList calls")
+	}
```
fake.tableListErrors = fake.tableListErrors[1:] is a pop, if len(fake.tableListErrors) == 0 then it is the same as `if calls > len(fake.tableListErrors)-1 `

aojea commented on pkg/proxy/conntrack/fake.go:
```diff
@@ -78,9 +81,45 @@ func (fake *fakeHandler) ConntrackDeleteFilters(tableType netlink.ConntrackTable
-	before := len(fake.entries)
-	for _, flow := range fake.entries {
+	// remainingFlows is the total flows remaining after simulating the deletion
+	var remainingFlows []*netlink.ConntrackFlow
+
+	var err error
+
+	calls := fake.callsConntrackDeleteFilters
+	fake.callsConntrackDeleteFilters++
+
+	// we ignore the cases when deleteErrors is not configured at all, this will happen when called
+	// from outside the current package (fake-proxiers), we don't simulate any error in this case.
+	if len(fake.deleteErrors) > 0 {
+		// we panic when ConntrackDeleteFilters() is called more times than the length of configured
+		// deleteErrors. This indicates improper configuration of fakeHandler for the test.
+		if calls > len(fake.deleteErrors)-1 {
+			panic("exceeded the allowed number of ConntrackDeleteFilters calls")
+		} else {
+			err = fake.deleteErrors[calls]
+		}
+	}
+
+	// the actual implementation netlink.ConntrackDeleteFilters does a dump call internally
+	// before deleting the flow entries, which can return a partial list of entries. We simulate
+	// the same behaviour here in our fake implementation.
+	if errors.Is(err, unix.EINTR) {
+		randInt := rand.IntN(len(fake.entries))
+		// simulate deletion on partial list in case of interrupt
+		flows = fake.entries[:randInt]
+		remainingFlows = fake.entries[randInt:]
+	} else if err != nil {
+		remainingFlows = fake.entries
+	} else {
```
The code of the `fake` is now more complex than the logic we are implementing, and it also has a lot of risks with arrays and indexes, how will this looks like just adding the retries?
do we really need this complexity?

adrianmoisey commented:
/close

Handled in https://github.com/kubernetes/kubernetes/pull/130256
